### ISSUE TITLE
Raises illegal memory access error randomly

### ISSUE THREAD

**Issue by zhangcx93** [JuliaGPU/CUDA.jl#1189]
**Describe the bug**

When running and testing my lib(with Pluto.jl), I found that it will raises `CUDA error: an illegal memory access was encountered (code 700, ERROR_ILLEGAL_ADDRESS)` sometimes. Most of the time it just works, no error.
the trace is something like(And repeated so many times in terminal):
```
throw_api_error(::CUDA.cudaError_enum)@error.jl:105
macro expansion@error.jl:115[inlined]
cuOccupancyMaxPotentialBlockSize(::Base.RefValue{Int32}, ::Base.RefValue{Int32}, ::CUDA.CuFunction, ::Ptr{Nothing}, ::Int64, ::Int64)@call.jl:26
#launch_configuration#39@occupancy.jl:59[inlined]
launch_configuration@occupancy.jl:56[inlined]
#launch_heuristic#233@gpuarrays.jl:22[inlined]
copyto!@broadcast.jl:65[inlined]
copyto!@broadcast.jl:936[inlined]
materialize!@broadcast.jl:894[inlined]
materialize!(::CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, ::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{1}, Nothing, typeof(*), Tuple{CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Float32}})@broadcast.jl:891
apply!(::Vector{CorticalSpikingNetwork.AbstractNeurons})@neurons.jl:159
```

This time the error line is just simply:
```
n.inhibitory_trace .*= 0.5f0
```
the type is just a normal CuArray

After this is happening, trying to run any CUDA code in the same session will 100% raise error, something with cuStreamCreate:
```
throw_api_error(::CUDA.cudaError_enum)@error.jl:105
macro expansion@error.jl:115[inlined]
cuStreamCreate(::Base.RefValue{Ptr{Nothing}}, ::CUDA.CUstream_flags_enum)@call.jl:26
(::CUDA.var"#CuStream#10#11")(::CUDA.CUstream_flags_enum, ::Nothing, ::Type{CUDA.CuStream})@stream.jl:21
CuStream@stream.jl:19[inlined]
stream@state.jl:374[inlined]
active_state@state.jl:107[inlined]
#_alloc#174@pool.jl:217[inlined]
#alloc#173@pool.jl:207[inlined]
alloc@pool.jl:203[inlined]
CuArray@array.jl:44[inlined]
CuArray@array.jl:125[inlined]
CuArray@array.jl:136[inlined]
(CUDA.CuArray{Float32, N, B} where {N, B})(::UndefInitializer, ::Int64)@array.jl:138
zeros(::Type, ::Int64)@array.jl:445
zeros@array.jl:447[inlined]
ArrayInput@inputs.jl:86[inlined]
ArrayInput@inputs.jl:85[inlined]
init_net()@Other: 2
init@Other: 4[inlined]
top-level scope@Local: 2[inlined]
```

**To reproduce**

I cannot find a stable way to reproduce this. Everytime when this happens, a simple restart of julia will fix it. But it comes back randomly.

<details><summary>Manifest.toml</summary>
<p>

```

[[CUDA]]
deps = ["AbstractFFTs", "Adapt", "BFloat16s", "CEnum", "CompilerSupportLibraries_jll", "ExprTools", "GPUArrays", "GPUCompiler", "LLVM", "LazyArtifacts", "Libdl", "LinearAlgebra", "Logging", "Printf", "Random", "Random123", "RandomNumbers", "Reexport", "Requires", "SparseArrays", "SpecialFunctions", "TimerOutputs"]
git-tree-sha1 = "335b3d2373733919b4972a51215a6840c7a33828"
uuid = "052768ef-5323-5732-b1bb-66c8b64840ba"
version = "3.4.2"

[[GPUArrays]]
deps = ["Adapt", "LinearAlgebra", "Printf", "Random", "Serialization", "Statistics"]
git-tree-sha1 = "69faa5f1c5706ca9ca067604acf797ee3a8ec6f6"
uuid = "0c68f7d7-f131-5f86-a1c3-88cf8149b2d7"
version = "8.1.1"

[[GPUCompiler]]
deps = ["ExprTools", "InteractiveUtils", "LLVM", "Libdl", "Logging", "TimerOutputs", "UUIDs"]
git-tree-sha1 = "4ed2616d5e656c8716736b64da86755467f26cf5"
uuid = "61eb1bfa-7361-4325-ad38-22787b887f55"
version = "0.12.9"
```

[[LLVM]]
deps = ["CEnum", "LLVMExtra_jll", "Libdl", "Printf", "Unicode"]
git-tree-sha1 = "46092047ca4edc10720ecab437c42283cd7c44f3"
uuid = "929cbde3-209d-540e-8aea-75f648917ca0"
version = "4.6.0"

</p>
</details>

**Expected behavior**

No Error

**Version info**

Details on Julia:

```
Julia Version 1.6.3
Commit ae8452a9e0 (2021-09-23 17:34 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
Environment:
  JULIA_GPU_ALLOWSCALAR = false
  JULIA_EDITOR = code
  JULIA_NUM_THREADS =
```

Details on CUDA:

```
CUDA toolkit 11.4.1, artifact installation
CUDA driver 11.4.0
NVIDIA driver 472.12.0

Libraries:
- CUBLAS: 11.5.4
- CURAND: 10.2.5
- CUFFT: 10.5.1
- CUSOLVER: 11.2.0
- CUSPARSE: 11.6.0
- CUPTI: 14.0.0
- NVML: 11.0.0+472.12
- CUDNN: 8.20.2 (for CUDA 11.4.0)
- CUTENSOR: 1.3.0 (for CUDA 11.2.0)

Toolchain:
- Julia: 1.6.3
- LLVM: 11.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device capability support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

1 device:
  0: NVIDIA GeForce RTX 2080 Ti (sm_75, 8.799 GiB / 11.000 GiB available)
```

**Additional context**

Running on windows with Pluto.jl, and start with `workspace_use_distributed=false`. I've been using like this for half a year, only starts to encounter this error from around last month.

**Comment by zhangcx93** [comment]
More stacktrace from terminal:
```
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\error.jl:105
  [2] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\error.jl:115 [inlined]
  [3] cuStreamCreate(phStream::Base.RefValue{Ptr{Nothing}}, Flags::CUDA.CUstream_flags_enum)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\utils\call.jl:26
  [4] CUDA.CuStream(; flags::CUDA.CUstream_flags_enum, priority::Nothing)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\stream.jl:21
  [5] CuStream
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\stream.jl:19 [inlined]
  [6] stream
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\state.jl:374 [inlined]
  [7] active_state
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\state.jl:107 [inlined]
  [8] #_free#177
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\pool.jl:293 [inlined]
  [9] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\pool.jl:278 [inlined]
 [10] macro expansion
    @ .\timing.jl:287 [inlined]
 [11] #free#176
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\pool.jl:277 [inlined]
 [12] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\array.jl:81 [inlined]
 [13] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\state.jl:175 [inlined]
 [14] unsafe_free!(xs::CUDA.CuArray{Int32, 1, CUDA.Mem.DeviceBuffer}, stream::CUDA.CuStream)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\array.jl:80
 [15] unsafe_finalize!(xs::CUDA.CuArray{Int32, 1, CUDA.Mem.DeviceBuffer})
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\array.jl:101
 [16] enable_finalizers
    @ .\gcutils.jl:123 [inlined]
 [17] unlock(l::Base.Threads.SpinLock)
    @ Base.Threads .\locks-mt.jl:92
 [18] unlock
    @ .\condition.jl:74 [inlined]
 [19] unlock(rl::ReentrantLock)
    @ Base .\lock.jl:139
 [20] lock(f::Pluto.var"#196#199"{Pluto.var"#153#158"{Pluto.ServerSession, Pluto.Notebook}}, l::ReentrantLock)
    @ Base .\lock.jl:189
 [21] flush
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:411 [inlined]
 [22] throttled_f
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:431 [inlined]
 [23] run_reactive!(session::Pluto.ServerSession, notebook::Pluto.Notebook, old_topology::Pluto.NotebookTopology, new_topology::Pluto.NotebookTopology, roots::Vector{Pluto.Cell}; deletion_hook::Function, persist_js_state::Bool, already_in_run::Bool)
    @ Pluto C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:105
 [24] run_reactive!
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:10 [inlined]
 [25] #162
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:138 [inlined]
 [26] macro expansion
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Tokens.jl:101 [inlined]
 [27] (::Pluto.var"#164#165"{Pluto.var"#162#163"{Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Pluto.ServerSession, Pluto.Notebook, Pluto.NotebookTopology, Pluto.NotebookTopology, Vector{Pluto.Cell}}})()
    @ Pluto .\task.jl:411
WARNING: Error while freeing DeviceBuffer(25.004 KiB at 0x000000020f75aa00):
CUDA.CuError(code=CUDA.cudaError_enum(0x000002bc), meta=nothing)
```

**Comment by maleadt** [comment]
These stacktraces are not useful, they invariably point to the first API call that detected the error, and are unrelated to the actual issue. Similarly, the fact that CUDA is broken afterwards is expected (that's how CUDA works).

WIthout a reproducer, I'm afraid this isn't actionable. Try outside of Pluto, running with `--check-bounds=yes`, maybe under `compute-sanizer`, etc.

**Comment by zhangcx93** [comment]
I managed to redirect the terminal output to a file to get the first trace:
```

Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\error.jl:105
  [2] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\error.jl:115 [inlined]
  [3] cuStreamCreate(phStream::Base.RefValue{Ptr{Nothing}}, Flags::CUDA.CUstream_flags_enum)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\utils\call.jl:26
  [4] CUDA.CuStream(; flags::CUDA.CUstream_flags_enum, priority::Nothing)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\stream.jl:21
  [5] CuStream
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\lib\cudadrv\stream.jl:19 [inlined]
  [6] stream
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\state.jl:374 [inlined]
  [7] active_state
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\state.jl:107 [inlined]
  [8] #_free#177
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\pool.jl:293 [inlined]
  [9] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\pool.jl:278 [inlined]
 [10] macro expansion
    @ .\timing.jl:287 [inlined]
 [11] #free#176
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\pool.jl:277 [inlined]
 [12] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\array.jl:81 [inlined]
 [13] macro expansion
    @ C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\state.jl:175 [inlined]
 [14] unsafe_free!(xs::CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, stream::CUDA.CuStream)
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\array.jl:80
 [15] unsafe_finalize!(xs::CUDA.CuArray{Float32, 1, CUDA.Mem.DeviceBuffer})
    @ CUDA C:\Users\zhang\.julia\packages\CUDA\9T5Sq\src\array.jl:101
 [16] show_richest(io::IOContext{IOBuffer}, x::Any)
    @ Pluto.PlutoRunner C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\runner\PlutoRunner.jl:777
 [17] sprint_withreturned(f::Function, args::Function; context::IOContext{Base.DevNull}, sizehint::Int64)
    @ Pluto.PlutoRunner C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\runner\PlutoRunner.jl:732
 [18] format_output_default(val::Any, context::Any)
    @ Pluto.PlutoRunner C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\runner\PlutoRunner.jl:656
 [19] #format_output#21
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\runner\PlutoRunner.jl:673 [inlined]
 [20] formatted_result_of(cell_id::Base.UUID, ends_with_semicolon::Bool, showmore::Nothing, workspace::Module)
    @ Pluto.PlutoRunner C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\runner\PlutoRunner.jl:589
 [21] top-level scope
    @ none:1
 [22] eval
    @ .\boot.jl:360 [inlined]
 [23] #137
    @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\remotecall.jl:354 [inlined]
 [24] run_work_thunk(thunk::Distributed.var"#137#138"{typeof(Core.eval), Tuple{Module, Expr}, Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, print_error::Bool)
    @ Distributed C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\process_messages.jl:63
 [25] #remotecall_fetch#142
    @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\remotecall.jl:379 [inlined]
 [26] remotecall_fetch
    @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\remotecall.jl:379 [inlined]
 [27] #remotecall_fetch#146
    @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\remotecall.jl:421 [inlined]
 [28] remotecall_fetch
    @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\remotecall.jl:421 [inlined]
 [29] remotecall_eval
    @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Distributed\src\macros.jl:246 [inlined]
 [30] (::Pluto.WorkspaceManager.var"#17#18"{Base.UUID, Bool, Nothing, Pluto.WorkspaceManager.Workspace})()
    @ Pluto.WorkspaceManager C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\WorkspaceManager.jl:290
 [31] withtoken(f::Pluto.WorkspaceManager.var"#17#18"{Base.UUID, Bool, Nothing, Pluto.WorkspaceManager.Workspace}, token::Pluto.Token)
    @ Pluto C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Tokens.jl:19
 [32] format_fetch_in_workspace(session_notebook::Pluto.WorkspaceManager.Workspace, cell_id::Base.UUID, ends_with_semicolon::Bool, showmore_id::Nothing)
    @ Pluto.WorkspaceManager C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\WorkspaceManager.jl:288
 [33] format_fetch_in_workspace
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\WorkspaceManager.jl:285 [inlined]
 [34] eval_format_fetch_in_workspace(session_notebook::Tuple{Pluto.ServerSession, Pluto.Notebook}, expr::Expr, cell_id::Base.UUID, ends_with_semicolon::Bool, function_wrapped_info::Nothing, contains_user_defined_macrocalls::Bool)
    @ Pluto.WorkspaceManager C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\WorkspaceManager.jl:271
 [35] run_single!(session_notebook::Tuple{Pluto.ServerSession, Pluto.Notebook}, cell::Pluto.Cell, reactive_node::Pluto.ReactiveNode, expr_cache::Pluto.ExprAnalysisCache; persist_js_state::Bool)
    @ Pluto C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:167
 [36] run_reactive!(session::Pluto.ServerSession, notebook::Pluto.Notebook, old_topology::Pluto.NotebookTopology, new_topology::Pluto.NotebookTopology, roots::Vector{Pluto.Cell}; deletion_hook::Function, persist_js_state::Bool, already_in_run::Bool)
    @ Pluto C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:110
 [37] run_reactive!
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:10 [inlined]
 [38] #162
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Run.jl:138 [inlined]
 [39] macro expansion
    @ C:\Users\zhang\.julia\packages\Pluto\nkHsc\src\evaluation\Tokens.jl:101 [inlined]
 [40] (::Pluto.var"#164#165"{Pluto.var"#162#163"{Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Pluto.ServerSession, Pluto.Notebook, Pluto.NotebookTopology, Pluto.NotebookTopology, Vector{Pluto.Cell}}})()
    @ Pluto .\task.jl:411
```

**Comment by maleadt** [comment]
> These stacktraces are not useful, they invariably point to the first API call that detected the error, and are unrelated to the actual issue.

See above. The illegal memory access comes from a kernel, not from an API call like that.

**Comment by zhangcx93** [comment]
I tried to use `--check-bounds=yes`, nothing is different.
Now it is stablely reproducable without Pluto.jl.
How can I use `compute-sanizer`? I tried `compute-sanizer.bat --launch-timeout=180 julia --project ./test.jl`, and nothing happens after
```========= COMPUTE-SANITIZER```
it just stalled.

If I don't use `--launch-timeout=180`, it just timeout after 10s.

**Comment by zhangcx93** [comment]
oh sorry I found my bug, i try to modify an sparse array after I changed the schema, so it get out of bounds. (I added a new function to modify its schema, no error in the new function alone, but other old function excuted after the new one is out of bounds, so it didn't come to my mind)
it was more random than it is today, and the trace log is not helpful. so I thought maybe it's a cuda bug.
perhaps a better tracelog or debugging tool is needed? why it doesnt just shows that line of code I wrote is having out-of-bound error..
